### PR TITLE
[Semi-Auto] LayerNorm Parallel Rule

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

New features
### PR changes

OPs
### Description

Pcard-70448

add spmd rule for layer_norm

### CODE REVIEW

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+#include "paddle/phi/core/distributed/auto_parallel/utils.h"
+
+namespace paddle {
+namespace distributed {
+namespace auto_parallel {
+using phi::distributed::auto_parallel::str_join;
+std::pair<std::vector<TensorDistAttr>, std::vector<TensorDistAttr>>
+LayerNormSPMDRule::InferForward(const std::vector<DistTensorSpec>& input_specs,
+                                const paddle::framework::AttributeMap& attrs) {
+  // step0: verify input args based on matmul logic
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc]: layer_norm

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    scale_axes = "k";
+    bias_axes = "k";
+  } else {
+    // z = x_axes.substr(begin_norm_axis, x_ndim - begin_norm_axis)
+    scale_axes = "z";
+    bias_axes = "z";
+  }
+  std::string out_axes = x_axes;
+
+  VLOG(4) << "LayerNormSPMDRule build Einsum notation (x,scale,bias->out): ["
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc:108]: should include all layer_norm outputs (x,scale,bias->out, mean, var)

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+  // step2.3: Merge and get Inputs' New Dims Mapping.
+  input_dist_attrs.emplace_back(x_dist_attr_dst);
+
+  // TODO(zhiqiu): support shardding on scale and bias
+  // Now, apply replicating.
+  input_dist_attrs.emplace_back(ReplicatedOnMesh(input_specs[1].dist_attr()));
+  input_dist_attrs.emplace_back(ReplicatedOnMesh(input_specs[2].dist_attr()));
+
+  // Step2.4.  handle input and out tensor partial
+  std::vector<int64_t> partial_on_dims;
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc]: LayerNorm activation output would not be partial

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+          << str_join(scale_shape) << "], src_dims_mapping: ["
+          << str_join(scale_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(input_dist_attrs[1].dims_mapping()) << "]; bias shape: ["
+          << str_join(bias_shape) << "], src_dims_mapping: ["
+          << str_join(bias_dims_mapping) << "], dst_dims_mapping: ["
+          << str_join(input_dist_attrs[2].dims_mapping())
+          << "]; out dims_mapping: [" << str_join(out_dims_mapping)
+          << "], partial_on_dims: [" << str_join(partial_on_dims) << "]";
+
+  return {input_dist_attrs, {output_dist_attr_dst}};
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc]: should infer the distattr of output （variance and mean）

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.h [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+#include <string>
+#include <vector>
+
+#include "paddle/fluid/distributed/auto_parallel/spmd_rules/common.h"
+
+namespace paddle {
+namespace distributed {
+namespace auto_parallel {
+
+TensorDistAttr GetInferedDistAttr(
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.h]: remove this

Comment (zhiqiu) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.h]: done

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc (hunk 2 above)
Comment (zhiqiu) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc:108]: done, thx

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc (hunk 1 above)
Comment (zhiqiu) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc]: done

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc (hunk 4 above)
Comment (zhiqiu) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc]: done thx

File: paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc (hunk 3 above)
Comment (zhiqiu) [reply paddle/fluid/distributed/auto_parallel/spmd_rules/layer_norm_spmd_rule.cc]: done, thx

File: paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc [hunk 6]
```diff
@@ -201,6 +199,101 @@ TEST(MatmulSPMDRule, Ctor) {
...
+      DistTensorSpec(scale_shape, scale_dist_attr);
+  DistTensorSpec bias_dist_tensor_spec =
+      DistTensorSpec(bias_shape, bias_dist_attr);
+
+  paddle::framework::AttributeMap attrs;
+  attrs["begin_norm_axis"] = 2;
+
+  SPMDRuleBase* layer_norm_rule = SPMDRuleMap::Instance().Get("layer_norm");
+
+  // ijk[1, -1, -1],k[-1],k[-1] --> ijk[1, -1, -1] partial[1]
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc:239]: no partial

File: paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc [hunk 7]
```diff
@@ -201,6 +199,101 @@ TEST(MatmulSPMDRule, Ctor) {
...
+  EXPECT_EQ(infered_dist_attrs.second[0].dims_mapping(),
+            std::vector<int64_t>({1, -1, -1}));
+  EXPECT_EQ(infered_dist_attrs.second[1].dims_mapping(),
+            std::vector<int64_t>({1}));
+  EXPECT_EQ(infered_dist_attrs.second[2].dims_mapping(),
+            std::vector<int64_t>({1}));
+  VLOG(4) << "test1 done.";
+
+  // ijk[1, 0, -1],k[0],k[0] --> ijk[1, 0, -1]
+  x_dist_tensor_spec.set_dims_mapping({1, 0, -1});
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc:265]: GOOD test case ! 
there are two kind of error in the case:
1. multiple batch axes (i,j) are sharded in input activation, which is not supported by now.
2. same mesh dimension ( 0 ) is sharding two different tensor axes (j & k).

it inspires me that we are missing one important precondition checking in InferForward (when support sharding on bias in future):
when [ijkl,y(kl),y(kl)->ijkl,x(ij),x(ij) (x,scale,bias->out,mean,variance, begin_norm_axis=2, x=ij, y=kl)]
thought "y" and "kl" are represented by different char, but they are same tensor axis and should be sharded by same mesh dim.

File: paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc [hunk 8]
```diff
@@ -201,6 +199,101 @@ TEST(MatmulSPMDRule, Ctor) {
...
+  scale_dist_tensor_spec.set_dims_mapping({0});
+  bias_dist_tensor_spec.set_dims_mapping({0});
+  EXPECT_ANY_THROW(
+      infered_dist_attrs = layer_norm_rule->InferForward(
+          {x_dist_tensor_spec, scale_dist_tensor_spec, bias_dist_tensor_spec},
+          attrs););
+  VLOG(4) << "test2 done.";
+
+  // ijk[0, -1, -1],z[-1],z[1] --> ijk[0, 1, -1, -1], z=jk
```
Comment (JZ-LIANG) [inline paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc:274]: ijk[0, -1, -1], k[-1], k[1] --> ijk[0, -1, -1], x[0], x[0], x=ij

**Review by JZ-LIANG — APPROVED** [review]
LGTM

File: paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc (hunk 6 above)
Comment (zhiqiu) [reply paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc:239]: I will fix it in the next pr.

File: paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc (hunk 8 above)
Comment (zhiqiu) [reply paddle/fluid/distributed/auto_parallel/test/spmd_rule_test.cc:274]: I will fix it in the next pr.

**Pull request merged**